File: day17/Day17.hs
{-# LANGUAGE TypeApplications #-}

import Control.Monad.State.Strict
import Data.Char as C
import Data.Function
import qualified Data.List as L

import System.Environment

data SimState = SimState
  { _position :: !(Int, Int)
  , _velocity :: !(Int, Int)
  , _maxY :: !Int
  , _hit :: !Bool
  }

type Area = (Int, Int, Int, Int)

simulate :: Area -> State SimState ()
simulate area@(xmin, xmax, ymin, ymax) = do
  (x, y) <- gets _position
  (xv, yv) <- gets _velocity
  maxY <- gets _maxY
  hit <- gets _hit
  put $
    SimState
      { _position = (x + xv, y + yv)
      , _velocity = (xv - signum xv, yv - 1)
      , _maxY = max maxY (y + yv)
      , _hit = hit || xmin <= x && x <= xmax && ymin <= y && y <= ymax
      }
  if y < ymin || x > xmax || (xv == 0 && x < xmin)
    then pure ()
    else simulate area

simulateRange :: [Int] -> [Int] -> Area -> (Int, Int)
simulateRange xrange yrange area =
  let results =
        [ (_maxY final, _hit final)
        | xv <- xrange
        , yv <- yrange
        , let final =
                execState
                  (simulate area)
                  SimState
                    { _position = (0, 0)
                    , _velocity = (xv, yv)
                    , _maxY = 0
                    , _hit = False
                    }
        , _hit final
        ]
   in (,) <$> maximum . map fst <*> length $ results

guess = 300
solveGen a@(xmin, xmax, ymin, ymax) =
  simulateRange [1 .. xmax + 1] [ymin - 1 .. guess] a

solve1 = fst . solveGen
solve2 = snd . solveGen

solutions = [solve1, solve2]

isIntChar = (||) <$> (== '-') <*> C.isNumber
numbers =
  map (read @Int)
    . filter (isIntChar . head)
    . L.groupBy ((==) `on` isIntChar)

main :: IO ()
main = do
  idx <- read . head <$> getArgs
  print
    . (solutions !! (idx - 1))
    . (\[a, b, c, d] -> (a, b, c, d))
    . numbers
    =<< getContents

File: day17/Day17.py
from sys import argv, stdin
import itertools as it

text = stdin.read().strip()

x1, x2, y1, y2 = area = [
    int("".join(v))
    for k, v in it.groupby(text, key=lambda s: s.isnumeric() or s == "-")
    if k
]


def in_area(x, y):
    x1, x2, y1, y2 = area

    return x1 <= x <= x2 and y1 <= y <= y2


def simulate(xv, yv):
    maxx = x = 0
    maxy = y = 0
    hit = in_area(x, y)

    while True:
        x += xv
        y += yv
        xv = xv + (-1 if xv > 0 else (1 if xv < 0 else 0))
        yv -= 1

        hit |= in_area(x, y)

        maxx = max(x, maxx)
        maxy = max(y, maxy)

        if y < y1 or x > x2 or (xv == 0 and x < x1):
            break

    return (hit, maxy)


count = 0
best = -1e9

# Pure guess
max_yv = 300

for xv in range(0, x2 + 1):
    for yv in range(y1 - 1, max_yv):
        hit, maxy = simulate(xv, yv)
        if hit:
            count += 1
            best = max(maxy, best)

if argv[1] == "1":
    print(best)
else:
    print(count)
